### ISSUE TITLE
[RFC] Custom editors

### ISSUE THREAD

**Issue by thorsteinsson** [aranja/tux#28]
- Start Date: 2017-02-28

# Summary

Use custom components to edit model fields.

# Motivation

Tux provides good default editors based on the data type. Developers can
replace the default ones or pick a specific editor component for a field in a model.

# Detailed design

Tux should export a function to register a custom editor. Default editors are provided by Tux, but all of them replaceable with custom ones.
Each editor has a React `component` to that renders the editor and `settings` that get passed to the component as props along with the model itself.
Editors can be applied to a data type, model and/or a field. If only `fieldId` is supplied it will be applied to all fields with that ID across all models, but if `modelId` is supplied as well, it's only for that one field.

```js
import { registerEditor } from 'tux'
import { MyCustomMarkdownEditor } from './components'

// Editor component for a data type globally
registerEditor({
  dataType: 'boolean',
  component: ToggleField,
})

// Editor component for email across all models
registerEditor({
  fieldId: 'email',
  component: TextField,
})

// Editor component for one field in one model
registerEditor({
  modelId: 'article'
  fieldId: 'description',
  component: MyCustomMarkdownEditor,
  settings: {
    helpText: 'Edit your content with style',
  },
})
```

The editor component gets the field as props.

```js
interface Props {
  field: {
    id: string
    value: string
    label: string
    onChange: () => void
  },
  helpText?: string, // Just a normal prop passed in as settings
}

class MyCustomMarkdownEditor extends Component<Props, any> {
  render() {
    const { label, value, onChange } = this.props.field
    return (
      <TextField label={label} value={value} onChange={onChange} />
    )
  }
}
```

# Alternatives

Contentful already offers [this](https://www.contentful.com/developers/docs/concepts/editor-interfaces/) but they use iframes, we get a much better UX using React components. It's interesting to check out their [UI Extensions](https://www.contentful.com/developers/docs/concepts/uiextensions/) for example usage of editor components.

# Unresolved questions

- Should editors have full access to the adapter?
- Should they have only get the field or the whole model?
- Should you be able to combine `fieldId`, `modelId` and `dataType` as you want or be more restrictive, like only a specfic field or a data type?

**Comment by eirikurn** [comment]
Sorry for the late review.

I imagined this more model driven. Let me brainstorm another api so we can discuss pros and cons.

```javascript
import { registerEditable, TextEditor, DateEditor, FileEditor } from 'tux'

// Set/overwrite schema for model with id 'article'.
registerEditable('article', [
  { field: 'fields.title', label: 'Title', component: TextEditor },

  // Props in object?
  {
    field: 'fields.description',
    label: 'Description',
    component: TextEditor,
    props: { multiLine: true },
  },

  // Props in new component?
  {
    field: 'fields.description',
    label: 'Description',
    component: (props) => <TextEditor {...props} placeholder="Hello world" />
  },

  { field: 'fields.publishedAt', label: 'Published At', component: DateEditor },
  { field: 'fields.image', label: 'Image', component: FileEditor },
])

// Edit schema?
registerEditable('article', schema =>
  schema.filter(field => field.field !== 'fields.publishedAt')
)
```

The use cases I'm trying to support are:

* Allow developer to hide fields from auto generated schema.
* Allow developer to tweak field props in auto generated schema.
* Allow developer to bring his own component for a field.
* Allow developer to overwrite a generated schema or define a schema when adapter doesn't generate it.
* Have a logical path from a given model to a schema to a set of fields.
* Make it "easy" for an adapter to auto-generate a schema.

What I'm not trying to support is:

* Allow developer to globally overwrite built-in editors/ui.
* Tux admin theming.

Additionally, I want minimal "magic" to give the developer a clearer mental image of why a specific editor appears or doesn't appear in the admin. As such, JSON dataTypes are quite limited (only string, number, boolean and array) and global mapping based on field name (for all models) could be quite confusing IMO.

*Regarding the unresolved questions.*
* These components should have full access to the tux context and by extension the adapter (this is needed for file uploads for example).
* They should accept `value` and `onChange` as a minimum (operating directly on the field value).

### Other thoughts
* Who manages the label and validation. I'd prefer tux to control the rendering of that but the component should definitely get some hooks, e.g. `id`, `isValid` and even a `onError` callback.
* I'm still trying to avoid forcing a common object structure for all adapters, but that means we need the adapter to do more "basic" things, like creating a valid file object, and figuring out an instance's model id.

**Comment by thorsteinsson** [comment]
In the initial draft the main thing was to override default editors, I agree that overriding per model is better. We'll tackling theming later down the road, I don't see any need for it now.

Regarding validation, it makes sense to have that independent of the editor. We could start off by reporting API errors correctly before implementing in-place validation.

I also like using `props` instead of `settings`, way clearer. I'll update the RFC.

**Comment by eirikurn** [comment]
Tracking RFC in aranja/tux-rfcs#1. Mostly implemented in #47
